kong/router/atc.lua

```diff
@@ -433,28 +455,78 @@ function _M:select(req_method, req_uri, req_host, req_scheme,
         return nil, err
       end
 
-    elseif req_headers and is_http_headers_field(field) then
+    elseif is_http_headers_field(field) then
+      if not req_headers then
+        goto continue
+      end
+
       local h = field:sub(14)
       local v = req_headers[h]
 
-      if v then
-        if type(v) == "string" then
-          local res, err = c:add_value(field, v:lower())
+      if not v then
+        goto continue
+      end
+
+      if type(v) == "string" then
+        local res, err = c:add_value(field, v:lower())
+        if not res then
+          return nil, err
+        end
+
+      else  -- type(v) == "table"
+        for idx = 1, #v do
+          local res, err = c:add_value(field, v[idx]:lower())
           if not res then
             return nil, err
           end
+        end
+      end -- if type(v)
+
+    elseif is_http_queries_field(field) then
+      if not req_queries then
+        goto continue
+      end
+
+      local n = field:sub(14)
+      local v = req_queries[n]
+
+      if not v then
+        goto continue
+      end
+
+      -- the query parameter has only one value, like /?foo=bar
+      if type(v) == "string" then
+        local res, err = c:add_value(field, v)
+        if not res then
+          return nil, err
+        end
 
-        else
-          for idx = 1, #v do
-            local res, err = c:add_value(field, v[idx]:lower())
-            if not res then
-              return nil, err
-            end
+      -- the query parameter has no value, like /?foo,
+      -- get_uri_arg will get a boolean `true`
+      -- we think it is equivalent to /?foo=
+      elseif type(v) == "boolean" then
+        local res, err = c:add_value(field, "")
+        if not res then
+          return nil, err
+        end
+
+      -- multiple values for a single query parameter, like /?foo=bar&foo=baz
+      else
+        for idx = 1, #v do
+          local res, err = c:add_value(field, v[idx])
+          if not res then
+            return nil, err
           end
         end
-      end
-    end
-  end
+      end -- type(v)
+
+    else  -- unknown field
+      error("unknown http schema field: " .. field, 2)
```

Comment:
I think we don't need `2` here, the error is not caused by the caller but the router code.

Reply:
done